### ISSUE TITLE
More type hints for Amazon provider

### ISSUE THREAD

**Issue by mik-laj** [apache/airflow#11297]
**Description**

The [`boto3`](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html) used by Amazon provider generates new classes at runtime, so mypy is not aware of types.  Luckily, there is a package that has generated type annotations, but we have to install them separately - [`boto3-stubs`](https://pypi.org/project/boto3-stubs/)
It would be nice for us to check if this library meets our expectations and integrate it with our CI. 

CC: @kaxil @mlgruby @coopergillan

**Use case / motivation**

This will allow us to detect more problems during the static check. It will also provide a better development experience as the IDE will be able to suggest types.

**Related Issues**

N/A

**Comment by mlgruby** [comment]
Since you mentioned in last issue comment I had a look and it make sense to try it here. 

I can have a look.

**Comment by kaxil** [comment]
> Since you mentioned in last issue comment I had a look and it make sense to try it here.
> 
> I can have a look.

Cool, thanks. Not super important, this can wait if you'd like to work on other Type hints first. Completely up to you :)

**Comment by mlgruby** [comment]
Yup, just gonna do some checks and then switch to other type hints 🙂

**Comment by scrambldchannel** [comment]
I'm interested in helping with this too, I was just looking at the AWS provider typehints and came across the same library.

**Comment by potix2** [comment]
I'm interested in this issue too.

**Comment by kaxil** [comment]
Go for it @potix2 if you would like to create PR, take https://github.com/apache/airflow/pull/11470 as an example

**Comment by potix2** [comment]
I'm working on #11531. But I feel it's too large to ask for review. Should I divide into some small requests?

**Comment by kaxil** [comment]
> I'm working on #11531. But I feel it's too large to ask for review. Should I divide into some small requests?

Sure got for it

**Comment by mik-laj** [comment]
@potix2 @mlgruby @scrambldchannel I assigned you to this ticket  🐱

**Comment by mik-laj** [comment]
@potix2 @mlgruby @scrambldchannel Do you need any help? If you have any questions about this ticket, I will be happy to explain.

**Comment by mlgruby** [comment]
I have started to work on this, sorry initially got distracted by professional work. I might shout out for help later. But for now I am good.

**Comment by potix2** [comment]
I'm working on #10742, but it's a little harder than I expected. After complete it, I'll be back to this issue.

**Comment by coopergillan** [comment]
@mik-laj - came back to this after seeing a new comment on an old PR: https://github.com/apache/airflow/pull/10164/files#r653685526

It looks like [`boto3_type_annotations`](https://github.com/alliefitter/boto3_type_annotations) is another option in addition to `boto3-stubs` as you mention.

Do you have a preference? Do you know if either of these have yet been installed? Guessing no.

**Comment by mik-laj** [comment]
boto3_type_annotations  is deprecated. We should use boto3-stubs.

**Comment by Taragolis** [comment]
I do not a time and desire to complete this (and other) tasks.
For someone who would like to work on this, implement [this one](https://github.com/apache/airflow/discussions/28560) could be nice start point
